## Importing the requirements

In [16]:
import numpy as np, pandas as pd # importing the requirements

In [17]:
election_dataset=pd.read_csv("election_results_2024.csv", encoding='utf-8-sig') # loading the dataset
election_dataset.shape # returns the number of rows and columns of the table

(543, 8)

## Inspecting and Exploring the data


In [18]:
election_dataset.head() # preprocessing the data

Constituency  Const. No.    Leading Candidate  \
0             AJMER          13  BHAGIRATH CHOUDHARY   
1             ALWAR           8      BHUPENDER YADAV   
2            AMBALA           1       VARUN CHAUDHRY   
3  ANANTNAG-RAJOURI           3     MIAN ALTAF AHMAD   
4         ARAKKONAM           7    S JAGATHRATCHAKAN   

                         Leading Party    Trailing Candidate  \
0               Bharatiya Janata Party  RAMCHANDRA CHOUDHARY   
1               Bharatiya Janata Party           LALIT YADAV   
2             Indian National Congress         BANTO KATARIA   
3  Jammu & Kashmir National Conference        MEHBOOBA MUFTI   
4            Dravida Munnetra Kazhagam             L VIJAYAN   

                             Trailing Party  Margin           Status  
0                  Indian National Congress  329991  Result Declared  
1                  Indian National Congress   48282  Result Declared  
2                    Bharatiya Janata Party   49036  Result Declared  
3  Jammu & Kashmir Peoples Democratic Party  281794  Result Declared  
4  All India Anna Dravida Munnetra Kazhagam  306559  Result Declared

In [19]:
election_dataset.tail()

Constituency  Const. No.         Leading Candidate  \
538            Wardha           8       AMAR SHARADRAO KALE   
539           Wayanad           4              RAHUL GANDHI   
540        West Delhi           6        KAMALJEET SEHRAWAT   
541  Yavatmal- Washim          14  SANJAY UTTAMRAO DESHMUKH   
542         Zahirabad           5      SURESH KUMAR SHETKAR   

                                        Leading Party  \
538  Nationalist Congress Party – Sharadchandra Pawar   
539                          Indian National Congress   
540                            Bharatiya Janata Party   
541             Shiv Sena (Uddhav Balasaheb Thackrey)   
542                          Indian National Congress   

                    Trailing Candidate            Trailing Party  Margin  \
538           RAMDAS CHANDRABHAN TADAS    Bharatiya Janata Party   81648   
539                         ANNIE RAJA  Communist Party of India  364422   
540                     MAHABAL MISHRA           Aam Aadmi Party  199013   
541  RAJSHRITAI HEMANT PATIL (MAHALLE)                 Shiv Sena   94473   
542                        B. B. PATIL    Bharatiya Janata Party   46188   

              Status  
538  Result Declared  
539  Result Declared  
540  Result Declared  
541  Result Declared  
542  Result Declared

In [20]:
election_dataset.describe()

Const. No.
count  543.000000
mean    18.178637
std     16.427840
min      1.000000
25%      6.000000
50%     14.000000
75%     25.000000
max     80.000000

In [21]:
election_dataset.describe(include="object").transpose()


count unique                       top freq
Constituency         543    541                Aurangabad    2
Leading Candidate    543    542              RAHUL GANDHI    2
Leading Party        543     42    Bharatiya Janata Party  240
Trailing Candidate   542    542      RAMCHANDRA CHOUDHARY    1
Trailing Party       542     42  Indian National Congress  167
Margin               543    543                    329991    1
Status               543      2           Result Declared  542

Observations:
* Aurangabad appeared twice which points towards the issue.
* Rahul Gandhi appeared twice because he has participated in two constituency.
* BJP is the leading party with a good control over 240 constituencies with 42 unique parties.
* INC appears to be trailing party in 167 constituencies.
* Margin is unique for each constituencies with 329991 as the highest.
* Result status have been declared for 542 constituencies, indicating the result of one constituency is still pending.

In [22]:
election_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 543 entries, 0 to 542
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Constituency        543 non-null    object
 1   Const. No.          543 non-null    int64 
 2   Leading Candidate   543 non-null    object
 3   Leading Party       543 non-null    object
 4   Trailing Candidate  542 non-null    object
 5   Trailing Party      542 non-null    object
 6   Margin              543 non-null    object
 7   Status              543 non-null    object
dtypes: int64(1), object(7)
memory usage: 34.1+ KB


* Trailing Candidate and Trailing Party contains the null values, Here it's clear that out of 543 entries 542 are non null and 1 is null value for the both of the columns.
* Margin column is of object type.

# Updating the encountered problems with the dataset


In [23]:
# converting the margin colum to numeric:
election_dataset["Margin"]=pd.to_numeric(election_dataset["Margin"], errors="coerce")

In [24]:
# filling the null values in the studied two columns
election_dataset = election_dataset.fillna({
    "Trailing Candidate":"N/A", 
    "Trailing Party":"N/A"
    })

In [25]:
election_dataset[election_dataset["Constituency"]=="Aurangabad"]

Constituency  Const. No.           Leading Candidate         Leading Party  \
41   Aurangabad          19  BHUMARE SANDIPANRAO ASARAM             Shiv Sena   
42   Aurangabad          37           ABHAY KUMAR SINHA  Rashtriya Janata Dal   

    Trailing Candidate                          Trailing Party    Margin  \
41  IMTIAZ JALEEL SYED  All India Majlis-E-Ittehadul Muslimeen  134650.0   
42  SUSHIL KUMAR SINGH                  Bharatiya Janata Party   79111.0   

             Status  
41  Result Declared  
42  Result Declared

Bhumare Asaram won lok sabha election at Aurangabad of Maharashtra
whereas Abhay Kumar won at Arurangabad of Bihar


In [26]:
election_dataset.loc[election_dataset["Constituency"]=="Aurangabad", "Constituency"]=["Aurangabad (Maharashtra)", "Aurangabad (Bihar)"]

In [27]:
# verifying the updated values
election_dataset[41:43]

Constituency  Const. No.           Leading Candidate  \
41  Aurangabad (Maharashtra)          19  BHUMARE SANDIPANRAO ASARAM   
42        Aurangabad (Bihar)          37           ABHAY KUMAR SINHA   

           Leading Party  Trailing Candidate  \
41             Shiv Sena  IMTIAZ JALEEL SYED   
42  Rashtriya Janata Dal  SUSHIL KUMAR SINGH   

                            Trailing Party    Margin           Status  
41  All India Majlis-E-Ittehadul Muslimeen  134650.0  Result Declared  
42                  Bharatiya Janata Party   79111.0  Result Declared

In [28]:
leading_party_counts=election_dataset["Leading Party"].value_counts()
leading_party_counts

Leading Party
Bharatiya Janata Party                                      240
Indian National Congress                                     99
Samajwadi Party                                              37
All India Trinamool Congress                                 29
Dravida Munnetra Kazhagam                                    22
Telugu Desam                                                 16
Janata Dal (United)                                          12
Shiv Sena (Uddhav Balasaheb Thackrey)                         9
Nationalist Congress Party – Sharadchandra Pawar              8
Independent                                                   7
Shiv Sena                                                     7
Lok Janshakti Party(Ram Vilas)                                5
Communist Party of India (Marxist)                            4
Yuvajana Sramika Rythu Congress Party                         4
Rashtriya Janata Dal                                          4
Aam Aadmi Party           

In [29]:
other_party_counts=leading_party_counts[leading_party_counts<5].sum()
other_party_counts

np.int64(52)

In [30]:
leading_party_counts = leading_party_counts[leading_party_counts>=5] # will allow only the entries of the parties having count greater than 5 
leading_party_counts["Others"]=other_party_counts # add a category into the leading party containing all the parties having count less than 5 
leading_party_counts

Leading Party
Bharatiya Janata Party                              240
Indian National Congress                             99
Samajwadi Party                                      37
All India Trinamool Congress                         29
Dravida Munnetra Kazhagam                            22
Telugu Desam                                         16
Janata Dal (United)                                  12
Shiv Sena (Uddhav Balasaheb Thackrey)                 9
Nationalist Congress Party – Sharadchandra Pawar      8
Independent                                           7
Shiv Sena                                             7
Lok Janshakti Party(Ram Vilas)                        5
Others                                               52
Name: count, dtype: int64

In [31]:
"Jai Shree Ram."

'Jai Shree Ram.'